File: parent_control.cpp
// parent_control.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <fstream>
#include <sstream>
#include <filesystem>
#include <ctime>
#include <json/json.h> // sudo apt-get install libjsoncpp-dev

using namespace std;
namespace fs = std::filesystem;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

string getConfigDir() {
    const char* home = getenv("HOME");
    if (!home) home = getenv("USERPROFILE");
    return string(home) + "/.parent_control";
}

string getConfigFile() {
    return getConfigDir() + "/config.json";
}

string getHostsFile() {
#ifdef _WIN32
    return "C:\\Windows\\System32\\drivers\\etc\\hosts";
#else
    return "/etc/hosts";
#endif
}

const string BLOCK_IP = "127.0.0.1";

Json::Value loadConfig() {
    fs::create_directories(getConfigDir());
    ifstream f(getConfigFile());
    Json::Value root;
    if (!f) {
        root["rules"]["sites"] = Json::arrayValue;
        root["rules"]["time_limit"] = 0;
        root["rules"]["apps"] = Json::arrayValue;
        root["active"] = false;
        root["start_time"] = Json::nullValue;
        root["total_time"] = 0;
        return root;
    }
    f >> root;
    return root;
}

void saveConfig(const Json::Value& root) {
    ofstream f(getConfigFile());
    f << root.toStyledString();
}

vector<string> readHosts() {
    ifstream f(getHostsFile());
    vector<string> lines;
    string line;
    while (getline(f, line)) lines.push_back(line);
    return lines;
}

void writeHosts(const vector<string>& lines) {
    ofstream f(getHostsFile());
    for (const auto& l : lines) f << l << endl;
}

void updateHosts(const vector<string>& domains, const string& action) {
    auto lines = readHosts();
    vector<string> newLines;
    for (const auto& line : lines) {
        bool skip = false;
        for (const auto& d : domains) {
            if (line.find(BLOCK_IP) == 0 && line.find(d) != string::npos) {
                skip = true;
                break;
            }
        }
        if (!skip) newLines.push_back(line);
    }
    if (action == "add") {
        for (const auto& d : domains) {
            newLines.push_back(BLOCK_IP + " " + d);
        }
    }
    writeHosts(newLines);
}

void addRule(const string& type, const string& value) {
    auto root = loadConfig();
    if (type == "site") {
        bool exists = false;
        for (const auto& s : root["rules"]["sites"]) {
            if (s.asString() == value) exists = true;
        }
        if (!exists) {
            root["rules"]["sites"].append(value);
            updateHosts({value}, "add");
        }
    } else if (type == "time") {
        int minutes = stoi(value);
        root["rules"]["time_limit"] = minutes;
    } else if (type == "app") {
        bool exists = false;
        for (const auto& a : root["rules"]["apps"]) {
            if (a.asString() == value) exists = true;
        }
        if (!exists) root["rules"]["apps"].append(value);
    } else {
        cout << colorize("Unknown rule type", RED) << endl;
        return;
    }
    saveConfig(root);
    cout << colorize("Rule added: " + type + " -> " + value, GREEN) << endl;
}

void removeRule(const string& type, const string& value) {
    auto root = loadConfig();
    if (type == "site") {
        Json::Value newSites = Json::arrayValue;
        for (const auto& s : root["rules"]["sites"]) {
            if (s.asString() != value) newSites.append(s);
        }
        root["rules"]["sites"] = newSites;
        updateHosts({value}, "remove");
    } else if (type == "time") {
        root["rules"]["time_limit"] = 0;
    } else if (type == "app") {
        Json::Value newApps = Json::arrayValue;
        for (const auto& a : root["rules"]["apps"]) {
            if (a.asString() != value) newApps.append(a);
        }
        root["rules"]["apps"] = newApps;
    } else {
        cout << colorize("Unknown rule type", RED) << endl;
        return;
    }
    saveConfig(root);
    cout << colorize("Rule removed: " + type + " -> " + value, GREEN) << endl;
}

void listRules() {
    auto root = loadConfig();
    cout << colorize("Current rules:", BLUE) << endl;
    if (root["rules"]["sites"].size() > 0) {
        cout << colorize("  Blocked sites:", YELLOW) << endl;
        for (const auto& s : root["rules"]["sites"])
            cout << "    - " << s.asString() << endl;
    }
    int timeLimit = root["rules"]["time_limit"].asInt();
    if (timeLimit > 0)
        cout << colorize("  Time limit: " + to_string(timeLimit) + " min", YELLOW) << endl;
    if (root["rules"]["apps"].size() > 0) {
        cout << colorize("  Blocked apps:", YELLOW) << endl;
        for (const auto& a : root["rules"]["apps"])
            cout << "    - " << a.asString() << endl;
    }
    if (root["rules"]["sites"].size() == 0 && timeLimit == 0 && root["rules"]["apps"].size() == 0)
        cout << colorize("  No rules.", YELLOW) << endl;
}

void enableControl() {
    auto root = loadConfig();
    if (root["active"].asBool()) {
        cout << colorize("Parent control already enabled.", YELLOW) << endl;
        return;
    }
    root["active"] = true;
    time_t now = time(nullptr);
    char buf[64];
    strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", localtime(&now));
    root["start_time"] = string(buf);
    saveConfig(root);
    // Apply site blocks
    vector<string> sites;
    for (const auto& s : root["rules"]["sites"]) sites.push_back(s.asString());
    if (!sites.empty()) updateHosts(sites, "add");
    cout << colorize("Parent control enabled.", GREEN) << endl;
}

void disableControl() {
    auto root = loadConfig();
    if (!root["active"].asBool()) {
        cout << colorize("Parent control already disabled.", YELLOW) << endl;
        return;
    }
    root["active"] = false;
    vector<string> sites;
    for (const auto& s : root["rules"]["sites"]) sites.push_back(s.asString());
    if (!sites.empty()) updateHosts(sites, "remove");
    saveConfig(root);
    cout << colorize("Parent control disabled.", GREEN) << endl;
}

void showStatus() {
    auto root = loadConfig();
    cout << colorize("Parent control: " + string(root["active"].asBool() ? "ENABLED" : "DISABLED"), BLUE) << endl;
    if (root["active"].asBool() && root.isMember("start_time") && !root["start_time"].isNull()) {
        string startStr = root["start_time"].asString();
        tm tm = {};
        stringstream ss(startStr);
        ss >> get_time(&tm, "%Y-%m-%dT%H:%M:%S");
        time_t start = mktime(&tm);
        int elapsed = (int)difftime(time(nullptr), start) / 60;
        cout << colorize("Session time: " + to_string(elapsed) + " min", YELLOW) << endl;
        int limit = root["rules"]["time_limit"].asInt();
        if (limit > 0) cout << colorize("Time limit: " + to_string(limit) + " min", YELLOW) << endl;
    }
    cout << colorize("Blocked sites: " + to_string(root["rules"]["sites"].size()), YELLOW) << endl;
    cout << colorize("Blocked apps: " + to_string(root["rules"]["apps"].size()), YELLOW) << endl;
}

void generateReport() {
    auto root = loadConfig();
    cout << colorize("Activity report:", BLUE) << endl;
    cout << "  Date: " << time(nullptr) << endl; // упрощённо
    cout << "  Total time: " << root["total_time"].asInt() << " min" << endl;
    cout << "  Time limit: " << root["rules"]["time_limit"].asInt() << " min" << endl;
    cout << "  Blocked sites: " << root["rules"]["sites"].size() << endl;
    cout << "  Blocked apps: " << root["rules"]["apps"].size() << endl;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cout << colorize("Usage: parent_control add|remove|list|enable|disable|status|report [args...]", YELLOW) << endl;
        return 1;
    }
    string cmd = argv[1];
    vector<string> args;
    for (int i = 2; i < argc; ++i) args.push_back(argv[i]);

    if (cmd == "add") {
        if (args.size() < 2) {
            cout << colorize("Usage: add <site|time|app> <value>", YELLOW) << endl;
            return 1;
        }
        addRule(args[0], args[1]);
    } else if (cmd == "remove") {
        if (args.size() < 2) {
            cout << colorize("Usage: remove <site|time|app> <value>", YELLOW) << endl;
            return 1;
        }
        removeRule(args[0], args[1]);
    } else if (cmd == "list") {
        listRules();
    } else if (cmd == "enable") {
        enableControl();
    } else if (cmd == "disable") {
        disableControl();
    } else if (cmd == "status") {
        showStatus();
    } else if (cmd == "report") {
        generateReport();
    } else {
        cout << colorize("Unknown command: " + cmd, RED) << endl;
    }
    return 0;
}

File: parent_control.go
// parent_control.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type Config struct {
	Rules struct {
		Sites     []string `json:"sites"`
		TimeLimit int      `json:"time_limit"`
		Apps      []string `json:"apps"`
	} `json:"rules"`
	Active    bool   `json:"active"`
	StartTime string `json:"start_time"`
	TotalTime int    `json:"total_time"`
}

var configDir = filepath.Join(os.Getenv("HOME"), ".parent_control")
var configFile = filepath.Join(configDir, "config.json")
var hostsFile = "/etc/hosts"
var blockIP = "127.0.0.1"

func init() {
	if runtime.GOOS == "windows" {
		hostsFile = `C:\Windows\System32\drivers\etc\hosts`
	}
}

func ensureConfig() {
	os.MkdirAll(configDir, 0755)
	if _, err := os.Stat(configFile); os.IsNotExist(err) {
		cfg := Config{}
		cfg.Rules.Sites = []string{}
		cfg.Rules.Apps = []string{}
		cfg.Active = false
		cfg.TotalTime = 0
		data, _ := json.MarshalIndent(cfg, "", "  ")
		ioutil.WriteFile(configFile, data, 0644)
	}
}

func loadConfig() Config {
	ensureConfig()
	data, _ := ioutil.ReadFile(configFile)
	var cfg Config
	json.Unmarshal(data, &cfg)
	return cfg
}

func saveConfig(cfg Config) {
	data, _ := json.MarshalIndent(cfg, "", "  ")
	ioutil.WriteFile(configFile, data, 0644)
}

func readHosts() []string {
	f, _ := os.Open(hostsFile)
	defer f.Close()
	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func writeHosts(lines []string) {
	ioutil.WriteFile(hostsFile, []byte(strings.Join(lines, "\n")), 0644)
}

func updateHosts(domains []string, action string) {
	lines := readHosts()
	var newLines []string
	for _, line := range lines {
		skip := false
		for _, d := range domains {
			if strings.TrimSpace(line)[:len(blockIP)] == blockIP && strings.Contains(line, d) {
				skip = true
				break
			}
		}
		if !skip {
			newLines = append(newLines, line)
		}
	}
	if action == "add" {
		for _, d := range domains {
			newLines = append(newLines, fmt.Sprintf("%s %s", blockIP, d))
		}
	}
	writeHosts(newLines)
}

func addRule(ruleType, value string) {
	cfg := loadConfig()
	switch ruleType {
	case "site":
		for _, s := range cfg.Rules.Sites {
			if s == value {
				fmt.Println(colorize("Site already in list", yellow))
				return
			}
		}
		cfg.Rules.Sites = append(cfg.Rules.Sites, value)
		updateHosts([]string{value}, "add")
	case "time":
		minutes, err := strconv.Atoi(value)
		if err != nil {
			fmt.Println(colorize("Invalid time value", red))
			return
		}
		cfg.Rules.TimeLimit = minutes
	case "app":
		for _, a := range cfg.Rules.Apps {
			if a == value {
				fmt.Println(colorize("App already in list", yellow))
				return
			}
		}
		cfg.Rules.Apps = append(cfg.Rules.Apps, value)
	default:
		fmt.Println(colorize("Unknown rule type", red))
		return
	}
	saveConfig(cfg)
	fmt.Println(colorize("Rule added: "+ruleType+" -> "+value, green))
}

func removeRule(ruleType, value string) {
	cfg := loadConfig()
	switch ruleType {
	case "site":
		newSites := []string{}
		for _, s := range cfg.Rules.Sites {
			if s != value {
				newSites = append(newSites, s)
			}
		}
		cfg.Rules.Sites = newSites
		updateHosts([]string{value}, "remove")
	case "time":
		cfg.Rules.TimeLimit = 0
	case "app":
		newApps := []string{}
		for _, a := range cfg.Rules.Apps {
			if a != value {
				newApps = append(newApps, a)
			}
		}
		cfg.Rules.Apps = newApps
	default:
		fmt.Println(colorize("Unknown rule type", red))
		return
	}
	saveConfig(cfg)
	fmt.Println(colorize("Rule removed: "+ruleType+" -> "+value, green))
}

func listRules() {
	cfg := loadConfig()
	fmt.Println(colorize("Current rules:", blue))
	if len(cfg.Rules.Sites) > 0 {
		fmt.Println(colorize("  Blocked sites:", yellow))
		for _, s := range cfg.Rules.Sites {
			fmt.Printf("    - %s\n", s)
		}
	}
	if cfg.Rules.TimeLimit > 0 {
		fmt.Println(colorize(fmt.Sprintf("  Time limit: %d min", cfg.Rules.TimeLimit), yellow))
	}
	if len(cfg.Rules.Apps) > 0 {
		fmt.Println(colorize("  Blocked apps:", yellow))
		for _, a := range cfg.Rules.Apps {
			fmt.Printf("    - %s\n", a)
		}
	}
	if len(cfg.Rules.Sites) == 0 && cfg.Rules.TimeLimit == 0 && len(cfg.Rules.Apps) == 0 {
		fmt.Println(colorize("  No rules.", yellow))
	}
}

func enableControl() {
	cfg := loadConfig()
	if cfg.Active {
		fmt.Println(colorize("Parent control already enabled.", yellow))
		return
	}
	cfg.Active = true
	cfg.StartTime = time.Now().Format(time.RFC3339)
	saveConfig(cfg)
	if len(cfg.Rules.Sites) > 0 {
		updateHosts(cfg.Rules.Sites, "add")
	}
	fmt.Println(colorize("Parent control enabled.", green))
}

func disableControl() {
	cfg := loadConfig()
	if !cfg.Active {
		fmt.Println(colorize("Parent control already disabled.", yellow))
		return
	}
	cfg.Active = false
	if len(cfg.Rules.Sites) > 0 {
		updateHosts(cfg.Rules.Sites, "remove")
	}
	saveConfig(cfg)
	fmt.Println(colorize("Parent control disabled.", green))
}

func showStatus() {
	cfg := loadConfig()
	status := "DISABLED"
	if cfg.Active {
		status = "ENABLED"
	}
	fmt.Println(colorize(fmt.Sprintf("Parent control: %s", status), blue))
	if cfg.Active && cfg.StartTime != "" {
		start, _ := time.Parse(time.RFC3339, cfg.StartTime)
		elapsed := int(time.Since(start).Minutes())
		fmt.Println(colorize(fmt.Sprintf("Session time: %d min", elapsed), yellow))
		if cfg.Rules.TimeLimit > 0 {
			fmt.Println(colorize(fmt.Sprintf("Time limit: %d min", cfg.Rules.TimeLimit), yellow))
		}
	}
	fmt.Println(colorize(fmt.Sprintf("Blocked sites: %d", len(cfg.Rules.Sites)), yellow))
	fmt.Println(colorize(fmt.Sprintf("Blocked apps: %d", len(cfg.Rules.Apps)), yellow))
}

func generateReport() {
	cfg := loadConfig()
	fmt.Println(colorize("Activity report:", blue))
	fmt.Printf("  Date: %s\n", time.Now().Format("2006-01-02"))
	fmt.Printf("  Total time: %d min\n", cfg.TotalTime)
	fmt.Printf("  Time limit: %d min\n", cfg.Rules.TimeLimit)
	fmt.Printf("  Blocked sites: %d\n", len(cfg.Rules.Sites))
	fmt.Printf("  Blocked apps: %d\n", len(cfg.Rules.Apps))
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println(colorize("Usage: parent_control add|remove|list|enable|disable|status|report [args...]", yellow))
		os.Exit(1)
	}
	cmd := os.Args[1]
	args := os.Args[2:]

	switch cmd {
	case "add":
		if len(args) < 2 {
			fmt.Println(colorize("Usage: add <site|time|app> <value>", yellow))
			return
		}
		addRule(args[0], args[1])
	case "remove":
		if len(args) < 2 {
			fmt.Println(colorize("Usage: remove <site|time|app> <value>", yellow))
			return
		}
		removeRule(args[0], args[1])
	case "list":
		listRules()
	case "enable":
		enableControl()
	case "disable":
		disableControl()
	case "status":
		showStatus()
	case "report":
		generateReport()
	default:
		fmt.Println(colorize("Unknown command: "+cmd, red))
	}
}
